application/src/main/java/org/togetherjava/tjbot/commands/moderation/WarnCommand.java

```diff
@@ -0,0 +1,120 @@
+package org.togetherjava.tjbot.commands.moderation;
+
+import net.dv8tion.jda.api.Permission;
+import net.dv8tion.jda.api.entities.Guild;
+import net.dv8tion.jda.api.entities.Member;
+import net.dv8tion.jda.api.entities.MessageEmbed;
+import net.dv8tion.jda.api.entities.User;
+import net.dv8tion.jda.api.events.interaction.SlashCommandEvent;
+import net.dv8tion.jda.api.interactions.Interaction;
+import net.dv8tion.jda.api.interactions.commands.OptionMapping;
+import net.dv8tion.jda.api.interactions.commands.OptionType;
+import net.dv8tion.jda.api.requests.RestAction;
+import net.dv8tion.jda.api.utils.Result;
+import org.jetbrains.annotations.NotNull;
+import org.togetherjava.tjbot.commands.SlashCommandAdapter;
+import org.togetherjava.tjbot.commands.SlashCommandVisibility;
+import org.togetherjava.tjbot.config.Config;
+
+import java.util.Objects;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+
+
+/**
+ * This command can warn users. The command will also try to DM the user to inform them about the
+ * action and the reason.
+ * <p>
+ * The command fails if the user triggering it is lacking permissions to either warn other users or
+ * to warn the specific given user (for example a moderator attempting to ban an admin).
+ */
+public final class WarnCommand extends SlashCommandAdapter {
+    private static final String USER_OPTION = "user";
+    private static final String REASON_OPTION = "reason";
+    private static final String ACTION_VERB = "warn";
+    private final ModerationActionsStore actionsStore;
+    private final Predicate<String> hasRequiredRole;
+
+    /**
+     * Creates a new Instance.
+     *
+     * @param actionsStore used to store actions issued by this command
+     */
+    public WarnCommand(@NotNull ModerationActionsStore actionsStore) {
+        super("warn", "warns the user", SlashCommandVisibility.GUILD);
```

Comment:
The description reads weird, also it should start with uppercase. Adjust it to `"Warns the given user"`.